Allow writing null valued keys in JSON

Jefffrey opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #3774

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

Allow users to write JSON with null values for keys, where previous (and default) is to skip the keys.

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

In arrow-json writer mod, new builder for `Writer` which allows configuring how null valued keys (and null arrays) should be handled. Will default to skipping keys with null values for backward compatibility.

# Are there any user-facing changes?

Yes

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

Jefffrey commented on arrow-json/src/writer.rs:
```diff
@@ -494,21 +575,36 @@ impl JsonFormat for JsonArray {
+/// [`SkipNulls`] for backward compatibility.
 #[derive(Debug)]
-pub struct Writer<W, F>
+pub struct Writer<W, F, N = SkipNulls>
 where
     W: Write,
     F: JsonFormat,
+    N: NullHandler,
```
I wasn't sure if it would be better to include this null handling logic in `JsonFormat` instead. Open to suggestions

Jefffrey replied:
Changed to be configured via a WriterBuilder now

tustvold commented:
As commented on the ticket, the current behaviour is simply incorrect as it will not survive a round-trip. I therefore don't think there is a need to make the prior logic configurable, which I think should simplify this PR

Jefffrey commented:
> As commented on the ticket, the current behaviour is simply incorrect as it will not survive a round-trip. I therefore don't think there is a need to make the prior logic configurable, which I think should simplify this PR

If you're referring to #5066 then that is a separate issue which this PR doesn't cover.

tustvold commented:
Aah I see, in that case is there a way we could make this an option instead of a generic. As it stands this is a breaking change, which I'm not sure about

Jefffrey commented:
Can you give an example where this might break existing uses? I thought including a default for the new generic parameter for `Writer` would make this backwards compatible,  but would be happy to be proven wrong.

Jefffrey commented:
> Regardless using generics here will result in additional codegen and I'm doubtful will improve performance. Adding a WriterOptions or similar should be more future proof and more consistent with other implementations

Yeah that makes 👍

I'll work on switching to such

Jefffrey commented:
Removed configuring this null behaviour as a generic trait, and changed to be a config that can be set via a new WriterBuilder (akin to arrow-csv writing behaviour)

tustvold reviewed (approved):
This looks good to me, my only comment is I think it should be `keep_null_values` or possibly `preserve_nulls`, as null keys are actually illegal in JSON :sweat_smile:

Jefffrey commented:
> This looks good to me, my only comment is I think it should be `keep_null_values` or possibly `preserve_nulls`, as null keys are actually illegal in JSON 😅

Yeah I wasn't sure on the proper name. I was gonna leave it as `keep_nulls` but I felt that could be confusing since it might imply affecting null values in arrays which should always be kept regardless.

Maybe something like `preserve_object_nulls`?

tustvold commented:
Perhaps explicit_nulls ?

Jefffrey commented:
> Perhaps explicit_nulls ?

Sounds better, renamed to that :+1:

tustvold reviewed (approved):
Looks good to me, thank you
